File: 1431-kids-with-the-greatest-number-of-candies/1431-kids-with-the-greatest-number-of-candies.cpp
int* maxElement(vector<int>::iterator begin, vector<int>::iterator end) {
    int* maxNum = NULL;
    for (vector<int>::iterator itr = begin; itr < end; itr++) {
        if (!maxNum || *maxNum < *itr)
            maxNum = &*itr;
    }
    
    return maxNum;
}
class Solution {
public:
    vector<bool> kidsWithCandies(vector<int>& candies, int extraCandies) {
        vector<bool> answer(candies.size());
        int* maxCandiesPtr = maxElement(candies.begin(), candies.end());
        
        for (int i = 0; i < candies.size(); i++) {
            answer[i] = candies[i] + extraCandies >= *maxCandiesPtr;
        }
        
        return answer;
    }
};

File: 1431-kids-with-the-greatest-number-of-candies/1431-kids-with-the-greatest-number-of-candies.rs
impl Solution {
    pub fn kids_with_candies(candies: Vec<i32>, extra_candies: i32) -> Vec<bool> {
        let max_candies = *match candies.iter().max() {
            Some(max) => max,
            None => panic!("Should never reach this")
        };
        let mut answer = Vec::with_capacity(candies.len());
        
        for candy in candies {
            answer.push(candy + extra_candies >= max_candies);
        }
        
        answer
    }
}